### PR TITLE
[SPARK-9358][SQL] Code generation for UnsafeRow joiner.

### PR DESCRIPTION
**Pull request by rxin**
This patch creates a code generated unsafe row concatenator that can be used to concatenate/join two UnsafeRows into a single UnsafeRow.

Since it is inherently hard to test these low level stuff, the test suites employ randomized testing heavily in order to guarantee correctness.

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 1]
```diff
@@ -69,12 +69,7 @@
    */
   public static boolean supportsAggregationBufferSchema(StructType schema) {
     for (StructField field: schema.fields()) {
-      if (field.dataType() instanceof DecimalType) {
-        DecimalType dt = (DecimalType) field.dataType();
-        if (dt.precision() > Decimal.MAX_LONG_DIGITS()) {
-          return false;
-        }
-      } else if (!UnsafeRow.settableFieldTypes.contains(field.dataType())) {
+      if (!UnsafeRow.settableFieldTypes.contains(field.dataType())) {
```
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: call isFixedLength ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+ */
+
+package org.apache.spark.sql.catalyst.expressions.codegen
+
+import org.apache.spark.sql.catalyst.expressions.{UnsafeRow, Attribute}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.unsafe.PlatformDependent
+
+
+abstract class UnsafeRowConcat {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala]: JoinUnsafeRow ?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java (hunk 1 above)
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: oops yes - missed that

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    val copyFixedLengthRow1 = s"""
+       |// Copy fixed length data for row1
+       |PlatformDependent.copyMemory(
+       |  obj1, offset1 + ${bitset1Words * 8},
+       |  buf, $cursor,
+       |  ${schema1.size * 8});
+     """.stripMargin
+
+    // --------------------- copy fixed length portion from row 2 ----------------------- //
+    cursor += schema1.size * 8
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala]: Should we move this into above section?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+       |  buf, $cursor,
+       |  numBytesVariableRow1);
+     """.stripMargin
+
+    // --------------------- copy variable length portion from row 2 ----------------------- //
+    val copyVariableLengthRow2 = s"""
+       |// Copy variable length data for row2
+       |long numBytesBitsetAndFixedRow2 = ${(bitset2Words + schema2.size) * 8};
+       |long numBytesVariableRow2 = row2.getSizeInBytes() - numBytesBitsetAndFixedRow2;
+       |PlatformDependent.copyMemory(
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala]: We could skip the copy if no variable length

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    // --------------------- copy bitset from row 1 ----------------------- //
+    val copyBitset1 = Seq.tabulate(bitset1Words) { i =>
+      s"""
+         |PlatformDependent.UNSAFE.putLong(buf, ${offset + i * 8},
+         |  PlatformDependent.UNSAFE.getLong(obj1, ${offset + i * 8}));
+       """.stripMargin
+    }.mkString
+
+
+    // --------------------- copy bitset from row 2 ----------------------- //
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateRowConcat.scala]: We could do something like this
```
Seq.tabulate(outputBitsetWords) { i => 
  val bitset = if (i< bitset1Words) {
    getLong(obj1, i * 8)
  } else if (i == bitset1Words && bitset1Remainder > 0) {
    getLong(obj1, i* 8) | getLong(obj2, 0) >>> bitset1Remainder
  } else {
    getLong(obj2, i - bitset1Words) <<< bitset1Remainder | getLong(obj2 i - bitset1Words + 1) >>> bitset1Remainder
  }
  putLong(buf, i * 8, bitset)
}
```

**Comment by rxin** [comment]
@davies I did everything except the following:
```
Seq.tabulate(outputBitsetWords) { i => 
  val bitset = if (i< bitset1Words) {
    getLong(obj1, i * 8)
  } else if (i == bitset1Words && bitset1Remainder > 0) {
    getLong(obj1, i* 8) | getLong(obj2, 0) >>> bitset1Remainder
  } else {
    getLong(obj2, i - bitset1Words) <<< bitset1Remainder | getLong(obj2 i - bitset1Words + 1) >>> bitset1Remainder
  }
  putLong(buf, i * 8, bitset)
}
```

I like the idea, but we can revisit this during QA to clean it up. Just have too many other things to do right now, and based on my experience, it takes a lot of time to get rewrite like this working. Filed a ticket to track it:  https://issues.apache.org/jira/browse/SPARK-9518

**Comment by davies** [comment]
LGTM

**Pull request closed without merging**